source/slang/slang-ir-any-value-inference.cpp

```diff
@@ -0,0 +1,231 @@
+                {
+                    _findDependenciesOfTypeInSet(field->getFieldType(), targetSet, result);
+                }
+            }
+            break;
+        default:
+            {
+                for (UInt i = 0; i < type->getOperandCount(); i++)
+                {
+                    if (auto operandType = as<IRType>(type->getOperand(i)))
+                        _findDependenciesOfTypeInSet(operandType, targetSet, result);
+                }
+            }
+            break;
+        }
+    }
+
+    List<IRInterfaceType*> findDependenciesOfTypeInSet(IRType* type, HashSet<IRInterfaceType*> targetSet)
+    {
+        List<IRInterfaceType*> result;
+        _findDependenciesOfTypeInSet(type, targetSet, result);
+
+        return result;
+    }
+
+    void _sortTopologically(
+        IRInterfaceType* interfaceType,
+        HashSet<IRInterfaceType*>& visited,
+        List<IRInterfaceType*>& sortedInterfaceTypes,
+        const Func<HashSet<IRInterfaceType*>, IRInterfaceType*>& getDependencies)
+    {
+        if (visited.contains(interfaceType))
+            return;
+
+        visited.add(interfaceType);
+
+        for (auto dependency : getDependencies(interfaceType))
+        {
+            _sortTopologically(dependency, visited, sortedInterfaceTypes, getDependencies);
+        }
+
+        sortedInterfaceTypes.add(interfaceType);
+    }
+    
+    List<IRInterfaceType*> sortTopologically(
+        HashSet<IRInterfaceType*> interfaceTypes,
+        const Func<HashSet<IRInterfaceType*>, IRInterfaceType*>& getDependencies)
+    {
+        List<IRInterfaceType*> sortedInterfaceTypes;
+        HashSet<IRInterfaceType*> visited;
+        for (auto interfaceType : interfaceTypes)
+        {
+            _sortTopologically(interfaceType, visited, sortedInterfaceTypes, getDependencies);
+        }
+        return sortedInterfaceTypes;
+    }
+
+    void inferAnyValueSizeWhereNecessary(
+        IRModule* module)
+    {
+        // Go through the global insts and collect all interface types.
+        // For each interface type, infer its any-value-size, by looking up
+        // all witness tables whose conformance type matches the interface type.
+        // then using _calcNaturalSizeAndAlignment to find the max size.
+        //
+        // Note: we only infer any-value-size for interface types that are used
+        // as a generic type parameter, because we don't want to infer any-value-size
+        // for interface types that are used as a witness table type.
+        //
+
+        HashSet<IRInst*> implementedInterfaces;
+        // Add all interface type that are implemented by at least one type to a set.
+        for (auto inst : module->getGlobalInsts())
+        {
+            if (inst->getOp() == kIROp_WitnessTable)
+            {
+                auto interfaceType = cast<IRWitnessTableType>(inst->getDataType())->getConformanceType();
+                implementedInterfaces.add(interfaceType);
```

Comment:
It doesn't appear to me that this step is very beneficial since we are going to go over the uses of an interface type later to collect all the implementations. If we found no implementation there we can just skip.